cpp/src/arrow/compute/kernels/scalar_temporal_unary.cc

```diff
@@ -168,8 +168,8 @@ struct TemporalComponentExtractWeek
 template <template <typename...> class Op, typename Duration, typename InType,
           typename OutType>
 struct TemporalComponentExtractRound
-    : public TemporalComponentExtractBase<Op, Duration, InType, OutType> {
-  using Base = TemporalComponentExtractBase<Op, Duration, InType, OutType>;
+    : public TemporalComponentExtractBase<Op, Duration, InType, InType> {
+  using Base = TemporalComponentExtractBase<Op, Duration, InType, InType>;
```

Comment:
This looks a little odd…it's because UnaryTemporalFactory is designed for functions which have a single consistent output type? May be worth noting down below in the registration section that the output type passed to the factory is ignored

Reply:
`OutputType(FirstType)` doesn't seem to resolve correctly. I tried making a new resolver but that didn't work either.
Added a comment now.

Reply:
I think the type resolver was working (else the test would have failed differently, the types would be wrong instead of the values) but UnaryTemporalFactory was passing the OutType onto the Exec template class which was using it for ScalarUnary: https://github.com/apache/arrow/blob/047202f5647e67dbc09c6080468675ebb7a78c44/cpp/src/arrow/compute/kernels/temporal_internal.h#L224
That in turn was probably truncating the values when writing to the output array: https://github.com/apache/arrow/blob/047202f5647e67dbc09c6080468675ebb7a78c44/cpp/src/arrow/compute/kernels/codegen_internal.h#L587

Reply:
Oh yeah indeed wrong output array type would be picked up.
I still can't trace down at what point truncation would happen though.

Reply:
I would guess in the `OutputAdapter` call since it's templated on the wrong type. Or actually when instantiating the kernel since it would be declared as `OutType Exec(...)`?